### ISSUE TITLE
Multiple simultaneous `uWS.App`s instances return a "truthy" `socket` on `listen` over the same port.

### ISSUE THREAD

**Issue by davidmurdoch** [uNetworking/uWebSockets.js#118]
uWebsockets.js version: v15.8.0

Minimal reproduction:
```
import uWS from "uWebSockets.js";
uWS.App({}).listen(8545, function(socket) {
  if (socket) {
    uWS.App({}).listen(8545, function(socket){
      if (socket) {
         // Didn't think it should get here but it does.
      }
    })
  }
})
```

I expected `listen` to fail causing the `socket` to be `false`.

System info:
* uWebSockets.js v15.8.0
* Ubuntu 18.04.2 LTS
* node.js v10.15.3

**Comment by alexhultman** [comment]
This is by design

**Comment by davidmurdoch** [comment]
Ah, ok. Seems unexpected for a node.js library.

**TL;DR:**

The behavior also happens across separate node _instances_.

I'm creating a node.js websocket-server CLI that utilizes uWS where users may want to start multiple instances simultaneously. Binding each CLI instance to the same port is always erroneous here. uWS doesn't currently offer a way to detect that it won't respond to certain routes over the specified port. It'd be great if it did!

Would a PR adding a flag to disable port sharing be something you'd be interested in?

---

**Long Version & Repro:**

*This part is more for passerbys who want to understand more about the behavior, than you, Alex, since you obviously already know how all this works. :-)*


The behavior also happens across separate node _instances_. Running the following code in two separate instances does not return any errors, `socket` is never `false`, and the second node instance will not be able to respond to any requests since the first one always handles them (as I'd expect), and I couldn't find any way to detect when a route will be ignored in the TS docs or C++ user manual.

```javascript
// run this code under two node instances simultaneously, then hit `http://localhost:8080`
const uWS = require("uWebSockets.js");
const randomId = (Math.random()*1000)|0;
uWS.App({})
  .any("/*", (response) => {
    response.end("response from: " + randomId);
  })
  .listen(8080, (socket) => {
    if (socket) {
      console.log("listening. Id: ", randomId);
    } else {
      throw new Error("Couldn't listen.");
    }
  });
```

I was hoping to use uWS in a CLI server application (https://www.npmjs.com/package/ganache-cli), where users often run multiple instances simultaneously, but they're supposed to run it on different ports. The cli _should_ return an error or fail if the port it needs to bind to is already bound (or the routes it requires are already being handled) -- a common mistake users make. Not being able to detect when a route can't be handled by a server complicates the implementation -- as it seems I'd need to create an out-of-process singleton that tracks port use somehow (I'm sure there are other ways of detecting this, too).

Would a PR adding a flag to disable port sharing be something you'd be interested in?

**Comment by alexhultman** [comment]
An option to listen exclusively to a port can be added but it's quite a bit of work and nobody else has expressed that interest.

**Comment by davidmurdoch** [comment]
My coworker, @seesemichaelj, put together a potential patch for this:

Changes:

https://github.com/seesemichaelj/uSockets/commit/f1c6a181f185bc541e9fdf001795d9ce90b3811f
https://github.com/seesemichaelj/uWebSockets/commit/ffd90db65ea3838235c5aa7620c6838bce4ca0a7
https://github.com/seesemichaelj/uWebSockets.js/commit/28b527960a95bcf25abf347463fa3c46d861f661

The changes are currently pretty minimal:

 * changed the `uSockets` library's `REUSE_PORT` flag to `DO_NOT_REUSE_PORT` and then adjusted the `if` statement.
 * added the relevant `listen()` functions in `uWebSockets` that add the optional `int options` argument.
 * added the logic in `uWebSockets.js` to decipher which C++ function the user was trying to access from its single `listen` function.

*Example use in uWebSockets.js:*
```javascript
import uWS from "uWebSockets.js";
const listenOptions = 2; // 2 is the flag for DO_NOT_REUSE_PORT
uWS.App({}).listen(8545, listenOptions, function(socket) {
  if (socket) {
    uWS.App({}).listen(8545, listenOptions, function(socket){
      if (!socket) {
         console.log("`socket` is falsy!");
      }
    })
  }
});
```

I've only tested on node.js v10.15.3 on Linux (Ubuntu 18.04.2 LTS) so far. Didn't want to to do too much work or update tests and documentation if this enhancement requires a different approach.

Let me know if this is a good approach or not.
```

**Comment by slavaGanzin** [comment]
@davidmurdoch Why don't you test if port is open with something like this: https://www.npmjs.com/package/is-port-reachable. You have minor inconvenience, so why you just "fix it" on your side? I think it's easier to write two lines in your code, than contribute 20 to someone's library.

**Comment by davidmurdoch** [comment]
@slavaGanzin For the same reason you shouldn't use `fs.exists` before accessing/writing a file: doing so introduces a race condition, since other processes may change the state between the two calls.

**Comment by alexhultman** [comment]
This change looks good, but I will probably change it a bit more before merging

**Comment by davidmurdoch** [comment]
Thanks, @alexhultman! I'll get the PRs out today.

**Comment by seesemichaelj** [comment]
The 3 related PRs can be seen referenced above. Of course @alexhultman, you'll need to update the submodule references with the final commit hashes.

Thanks a lot!

**Comment by alexhultman** [comment]
Okay it should now be LIBUS_LISTEN_EXCLUSIVE_PORT

**Comment by KrishnaPG** [comment]
How to specify this `LIBUS_LISTEN_EXCLUSIVE_PORT` option when both `hostname` and `port` are used? 

Currently `ListenOptions` is allowed in only in one particular case, and it does not allow specifying the hostname: 
https://github.com/uNetworking/uWebSockets.js/blob/d0f292446a7a73802df14257cafb0d4b52540b94/docs/index.d.ts#L268-L272

**Comment by PMA365** [comment]
@KrishnaPG 
> How to specify this `LIBUS_LISTEN_EXCLUSIVE_PORT` option when both `host` and `port` are used?

Add these line to your index.d.ts file inside `export interface TemplatedApp { }` in your Project :
```javascript
// my custom Listen port
    /** Listens to hostname & port and sets Listen Options. Callback hands either false or a listen socket. */
    listen(host: RecognizedString, port: number,options: ListenOptions, cb: (listenSocket: us_listen_socket | false) => void | Promise<void>) : TemplatedApp;
```
I have tested and it's working. When you try to listen to same port for the second time it can't listen and will print : Failed to listen to port  (yourPort)

**Comment by KrishnaPG** [comment]
Great - thanks @PMA365  Hope this will be pulled into the main

**Comment by uNetworkingAB** [comment]
Make a pr of it
